src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs

```diff
@@ -3272,42 +3271,40 @@ private void GetRPCObject(int paramCount, ref _SqlRPC rpc)
                 rpc = _rpcArrayOf1[0];
             }
 
-            rpc.ProcID = 0;
-            rpc.rpcName = null;
             rpc.options = 0;
+            rpc.systemParamCount = systemParamCount;
 
-
+            int currentCount = rpc.systemParams?.Length ?? 0;
             // Make sure there is enough space in the parameters and paramoptions arrays
-            if (rpc.parameters == null || rpc.parameters.Length < paramCount)
-            {
-                rpc.parameters = new SqlParameter[paramCount];
-            }
-            else if (rpc.parameters.Length > paramCount)
+            if (currentCount < systemParamCount)
             {
-                rpc.parameters[paramCount] = null;    // Terminator
+                Array.Resize(ref rpc.systemParams, systemParamCount);
+                Array.Resize(ref rpc.systemParamOptions, systemParamCount);
+                for (int index = currentCount; index < systemParamCount; index++)
+                {
+                    rpc.systemParams[index] = new SqlParameter();
+                }
             }
-            if (rpc.paramoptions == null || (rpc.paramoptions.Length < paramCount))
+            for (int ii = 0; ii < systemParamCount; ii++)
             {
-                rpc.paramoptions = new byte[paramCount];
+                rpc.systemParamOptions[ii] = 0;
             }
-            else
+            if ((rpc.userParamMap?.Length ?? 0) < userParamCount)
             {
-                for (ii = 0; ii < paramCount; ii++)
-                    rpc.paramoptions[ii] = 0;
+                Array.Resize(ref rpc.userParamMap, userParamCount);
             }
         }
 
+
         private void SetUpRPCParameters(_SqlRPC rpc, int startCount, bool inSchema, SqlParameterCollection parameters)
```

Comment:
startCount parameter is no longer in use. Can it be removed?

Reply:
yup, done.